From 0cc22ec515e825d9ad721c6308d34fd5b38c9223 Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentinx.casasnovas@intel.com>
Date: Wed, 14 Nov 2012 20:26:43 +0100
Subject: [PATCH 53/78] wl12xx: Fix memleak in wl1271_acx_init_mem_config

BZ: 68066

The function wl1271_acx_init_mem_config did not check whether or not the
wl->target_mem_map field was previsouly allocated. In the case it was, it would
cause a memory leak by overwriting the first one allocated.

Change-Id: I61591cdb4861c149a4d9710db32cc886c49733ea
Signed-off-by: Quentin Casasnovas <quentinx.casasnovas@intel.com>
Reviewed-on: http://android.intel.com:8080/74102
Reviewed-by: Regairaz, Julie <julie.regairaz@intel.com>
Reviewed-by: De Cesco, Jonathan <jonathan.de.cesco@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
Reviewed-on: http://android.intel.com:8080/78010
Reviewed-by: Li, XiaojingX <xiaojingx.li@intel.com>
Tested-by: Li, XiaojingX <xiaojingx.li@intel.com>
---
 wl12xx-compat/drivers/net/wireless/wl12xx/acx.c  |    7 +++++--
 wl12xx-compat/drivers/net/wireless/wl12xx/main.c |    1 +
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/acx.c b/wl12xx-compat/drivers/net/wireless/wl12xx/acx.c
index b99d79f..5c3c228 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/acx.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/acx.c
@@ -1028,8 +1028,11 @@ int wl1271_acx_init_mem_config(struct wl1271 *wl)
 {
 	int ret;
 
-	wl->target_mem_map = kzalloc(sizeof(struct wl1271_acx_mem_map),
-				     GFP_KERNEL);
+	if (wl->target_mem_map)
+		memset(wl->target_mem_map, 0, sizeof(struct wl1271_acx_mem_map));
+	else
+		wl->target_mem_map = kzalloc(sizeof(struct wl1271_acx_mem_map),
+					     GFP_KERNEL);
 	if (!wl->target_mem_map) {
 		wl1271_error("couldn't allocate target memory map");
 		return -ENOMEM;
diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
index 7ba4882..7555577 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
@@ -6318,6 +6318,7 @@ static struct ieee80211_hw *wl1271_alloc_hw(void)
 	wl->system_hlid = WL12XX_SYSTEM_HLID;
 	wl->active_sta_count = 0;
 	wl->fwlog_size = 0;
+	wl->target_mem_map = NULL;
 	init_waitqueue_head(&wl->fwlog_waitq);
 
 	/* The system link is always allocated */
-- 
1.7.9

